sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,711 @@
+
+    @Override
+    public UnboundedSource.CheckpointMark getCheckpointMark() {
+      return checkpointMark;
+    }
+
+    @Override
+    public RabbitMQSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public byte[] getCurrentRecordId() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      if (currentRecordId != null) {
+        return currentRecordId;
+      } else {
+        return "".getBytes();
+      }
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (currentTimestamp == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+    @Override
+    public RabbitMqMessage getCurrent() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      ConnectionFactory connectionFactory;
+      try {
+        connectionFactory =
+            source.spec.connectionConfig().createConnectionFactory();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      try {
+        connection = connectionFactory.newConnection();
+        channel = connection.createChannel();
+        if (channel == null) {
+          throw new IOException("No RabbitMQ channel available");
+        }
+        String queueName = source.spec.queue();
+        if (source.spec.queueDeclare()) {
+          channel.queueDeclare(queueName, false, false, false, null);
+        }
+        if (source.spec.exchange() != null) {
+          channel.exchangeDeclare(source.spec.exchange(), source.spec.exchangeType());
+          queueName = channel.queueDeclare().getQueue();
+          channel.queueBind(queueName, source.spec.exchange(), source.spec.routingKey());
+        }
+        checkpointMark.channel = channel;
+        consumer = new QueueingConsumer(channel);
+        channel.txSelect();
+        channel.basicConsume(queueName, false, consumer);
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      try {
+        QueueingConsumer.Delivery delivery = consumer.nextDelivery();
+        if (source.spec.useCorrelationId()) {
+          String correlationId = delivery.getProperties().getCorrelationId();
+          if (correlationId == null) {
+            throw new IOException("RabbitMqIO.Read uses message correlation ID, but received "
+                + "message has a null correlation ID");
+          }
+          currentRecordId = correlationId.getBytes();
+        }
+        long deliveryTag = delivery.getEnvelope().getDeliveryTag();
+        checkpointMark.sessionIds.add(deliveryTag);
+        RabbitMqMessage message = new RabbitMqMessage(
+            source.spec.routingKey(),
+            delivery.getBody(),
+            delivery.getProperties().getContentType(),
```

Comment:
This looks like a lot of boilerplate, and it's brittle - a new version of AMQP will make this code silently omit some data. Why not just pass `delivery` itself to `RabbitMqMessage`?

If the problem is that it's not serializable, then I think we should treat this the same way we did with JdbcIO.read, where ResultSet is not serializable - and make the IO return not a RabbitMqMessage, but instead a user's type, and take a `SerializableFunction<Delivery, T>`. Likewise for the write().

Reply:
I'm just wrapping the data we actually need in the IO itself. I think it's enough as a first version of the IO.